hare/algorithm.go

```diff
@@ -0,0 +1,270 @@
+}
+
+func NewConsensusProcess(key PubKey, layer LayerId, s Set, oracle Rolacle, signing Signing, p2p NetworkService, broker *Broker) *ConsensusProcess {
+	proc := &ConsensusProcess{}
+	proc.State = State{0, 0, s, Certificate{}}
+	proc.Closer = NewCloser()
+	proc.pubKey = key
+	proc.layerId = layer
+	proc.oracle = oracle
+	proc.signing = signing
+	proc.network = p2p
+	proc.inbox = broker.CreateInbox(&layer)
+	proc.knowledge = make([]*pb.HareMessage, 0)
+	proc.isProcessed = make(map[uint32]bool)
+
+	return proc
+}
+
+func (proc *ConsensusProcess) Start() error {
+	if !proc.startTime.IsZero() { // called twice on same instance
+		log.Error("ConsensusProcess has already been started.")
+		return errors.New("failed starting consensus process")
+	}
+
+	proc.startTime = time.Now()
+
+	// send first status message
+	m, err := proc.buildStatusMessage()
+	if err != nil {
+		return errors.New("failed starting consensus process")
+	}
+
+	proc.network.Broadcast(ProtoName, m)
+
+	go proc.eventLoop()
+
+	return nil
+}
+
+func (proc *ConsensusProcess) WaitForCompletion() {
+	select {
+	case <-proc.CloseChannel():
+		// TODO: access broker and remove layer (consider register/unregister terminology)
+		return
+	}
+}
+
+func (proc *ConsensusProcess) eventLoop() {
+	log.Info("Start listening")
+
+	ticker := time.NewTicker(RoundDuration)
+
+	for {
+		select {
+		case msg := <-proc.inbox: // msg event
+			proc.handleMessage(msg)
+		case <-ticker.C: // next round event
+			proc.nextRound()
+		case <-proc.CloseChannel(): // close event
+			log.Info("Stop event loop, instance aborted")
+			return
+		}
+	}
+}
+
+func (proc *ConsensusProcess) handleMessage(m *pb.HareMessage) {
+	// Note: layer is already verified by the broker
+
+	// verify iteration
+	if m.Message.K != proc.k {
+		return
+	}
+
+	// validate signature
+	data, err := proto.Marshal(m.Message)
+	if err != nil {
+		return
+	}
+	if !proc.signing.Validate(data, m.InnerSig) {
+		return
+	}
+
+	// validate role
+	if !proc.oracle.ValidateRole(roleFromIteration(m.Message.K),
+		RoleRequest{PubKey{NewBytes32(m.PubKey)}, LayerId{NewBytes32(m.Message.Layer)}, m.Message.K},
+		Signature(m.Message.RoleProof)) {
+		return
+	}
+
+	proc.knowledge = append(proc.knowledge, m)
+}
+
+func (proc *ConsensusProcess) nextRound() {
+	log.Info("End of round: %d", proc.k)
+
+	// process the round
+	// state is passed by value, ownership of knowledge is also passed
+	go proc.processMessages(proc.State, proc.knowledge)
+
+	// advance to next round
+	proc.k++
+
+	// can reset knowledge as ownership was passed before
+	proc.knowledge = make([]*pb.HareMessage, InitialKnowledgeSize)
+}
+
+// TODO: put state and msg in Context struct?
+func (proc *ConsensusProcess) processMessages(state State, msgs []*pb.HareMessage) {
+	// only one process should run concurrently
+	proc.mutex.Lock()
+	defer proc.mutex.Unlock()
+
+	// check if process of messages has already been made for this round
+	if _, exist := proc.isProcessed[state.k]; exist {
```

Comment:
How can it happen? If it shouldn't happen that this check basically hides bugs

Reply:
If we support early sending then it can happen. We could have an early process of the messages and be done with it, but the next "tick" should know about it.